### PR TITLE
added formData to validation function

### PR DESCRIPTION
**Pull request by alexrybch**
Signed-off-by: Alex Rybchenko <arybchenko@box.com>

## Hey, I just made a Pull Request!

I think it could be useful to have `formData` in validation function to perform async validation on the fields that depends on value of other fields. 

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ export const createAsyncValidators = (
         if (validator) {
           const fieldValidation = createFieldValidation();
           try {
-            await validator(value, fieldValidation, context);
+            await validator(value, fieldValidation, { ...context, formData });
```
Comment (camilaibs) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts:45]: In terms of documentation, could you add a use case example [here](https://github.com/backstage/backstage/blob/master/docs/features/software-templates/writing-custom-field-extensions.md) in the official docs as well as [TSDoc @example](https://tsdoc.org/pages/tags/example/) to update the [API reference page](https://backstage.io/docs/reference/plugin-scaffolder.customfieldvalidator)?

There's a note discouraging relying on formData [here](https://backstage.io/docs/features/software-templates/writing-custom-field-extensions#access-data-from-other-fields). That consideration seems to be valid for validators as well, so we could mention same note in these new examples as well.

**Review by camilaibs — COMMENTED** [review]
Thank you for contributing 🙌🏻 , I would like the maintainer's point of view about the proposal. Other than that, I added only a note about documentation 🙂

**Comment by camilaibs** [comment]
> Thank you for contributing 🙌🏻 , I would like the maintainer's point of view about the proposal

What do you think about the proposal @benjdlambert ?

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts (hunk 1 above)
Comment (alexrybch) [reply plugins/scaffolder/src/next/TemplateWizardPage/Stepper/createAsyncValidators.ts:45]: Sure, will add those

**Comment by Rugvip** [comment]
@alexrybch please sign your commits with DCO so that we can move this forward

**Review by Rugvip — CHANGES_REQUESTED** [review]
Could you add a changeset as well? 🙏 

Implementation looks alright but gonna leave whether this is the direction we want to go to @benjdlambert

**Comment by benjdlambert** [comment]
Hey sorry! Had this on my list for a while, just getting round to this.

I think it's important that we address the fact that whilst it's possible, it's not really encouraged in the docs like we do in with the `formData` in the `formContext` field.

**Review by Rugvip — DISMISSED** [review]
Thanks! 👍 

Let's add this, but as Camila and Ben pointed out I think it's best to add a section to the docs that talks about the `context` parameter but also discourages the use of `formData`.

**Comment by alexrybch** [comment]
> Oh sorry, didn't realize this was just for the Next fields. In that case there's no need to make this a `minor` since it's `alpha` API, we can call it a `patch` and there's no need to document yet, so never mind that.

Changed to patch and added discouraging text to changeset :)

**Comment by benjdlambert** [comment]
@alexrybch The changeset's lifespan is pretty short lived, so if we want people to see this information I'd suggest that we put it in the documentation somewhere around the previous callout https://backstage.io/docs/features/software-templates/writing-custom-field-extensions#access-data-from-other-fields

**Comment by alexrybch** [comment]
> @alexrybch The changeset's lifespan is pretty short lived, so if we want people to see this information I'd suggest that we put it in the documentation somewhere around the previous callout https://backstage.io/docs/features/software-templates/writing-custom-field-extensions#access-data-from-other-fields

The change is only in "next" scaffolder version which is not in the docs yet. Maybe it's better to wait until it becomes a default?

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**